src/axom/core/ArrayBase.hpp

```diff
@@ -477,6 +532,74 @@ bool allNonNegative(const T (&arr)[N])
   return true;
 }
 
+/// \brief Indirection needed to dodge an MSVC compiler bug
+template <typename... Args>
+struct all_types_are_integral_impl : std::true_type
+{ };
+
+template <typename First, typename... Rest>
+struct all_types_are_integral_impl<First, Rest...>
+{
+  static constexpr bool value = std::is_integral<First>::value &&
+    all_types_are_integral_impl<Rest...>::value;
+};
+
+/// \brief Checks if the first type in a parameter pack is integral
+template <typename... Args>
+struct all_types_are_integral
+{
+  static constexpr bool value = all_types_are_integral_impl<Args...>::value;
+};
+
+/// \brief Translates between the MemorySpace enum and Umpire allocator IDs
+template <MemorySpace SPACE>
+inline int getAllocatorID();
+
```

Comment:
Minor (optional): It might be useful to have a utility function of the form:  
```cpp
bool areCompatible(MemorySpace space1, MemorySpace space2);
```
And possibly also:
```cpp
#ifdef AXOM_USE_UMPIRE
bool areCompatible(MemorySpace SPACE, int allocatorID);
#endif
```

Reply:
I can understand the utility of the second function (in particular, relating to @kanye-quest's comment about pooled allocators and also converting from Dynamic -> another space) but I'm not sure where/how the first one would be used.